reps/2022-12-15-optimizer-data.md

```diff
@@ -0,0 +1,308 @@
+
+Lazy mode has proved to be effective and is the de-facto industry standard in the big data processing world (Spark, Presto/Trino, Hive). The major motivation with lazy mode is to build a plan optimizer to generate the optimal execution plan. It opens the door for more advanced techniques such as adaptive query execution, cost-based optimizer, pluggable execution back-end, etc. Lazy mode is the first step of the execution optimizer. Execution optimizer is the route to get us to optimal performance.
+
+### 2. Introduce execution plan optimizer
+
+#### 2.1. How do things work today?
+
+* `Dataset` has `ExecutionPlan`. 
+* `ExecutionPlan` has `BlockList` (as input) and a list of `Stage`s (as execution). 
+* When executing `ExecutionPlan`, `ExecutionPlan._optimize()` is called to get an optimal list of `Stage`s. 
+* After `_optimize()` is done, each `Stage` is executed one by one sequentially. First `Stage` reads input `BlockList`. All other `Stage` reads the output `BlockList` from the previous `Stage`, which is materialized in the Ray Object Store. 
+* `ExecutionPlan` supports partial execution (so-called snapshot), i.e. only executing some of `Stage`s but not all.
+
+#### 2.2. Proposed Change
+
+The current design works for now, but has a lot of room for improvement. Without a proper plan/query optimizer, it’s hard to extend and engineeringly impossible to maintain cleanly.
+
+##### 2.2.1. Interfaces
+
+```python
+class Optimizer():
+  """Abstract class for optimizer.
+
+  An optimizer transforms the plan, with a list of optimization rules and input statistics.
+  """
+
+  rules: List[Rule]
+
+  def optimize(plan: Plan, statistics) -> Plan:
+    for rule in rules:
+      plan = rule.apply(plan, statistics)
+    return plan 
+
+class LogicalOptimizer(Optimizer):
+  """Optimizer for logical plan.
+
+  An optimizer transforms the logical plan, with a list of optimization Rules and input statistics.
+  """
+
+  rules: List[Rule]
+
+  def optimize(plan: LogicalPlan, statistics) -> LogicalPlan:
+    for rule in rules:
+      plan = rule.apply(plan, statistics)
+    return plan 
+
+class PhysicalOptimizer(Optimizer):
+  """Optimizer for physical plan.
+
+  An optimizer transforms the physical plan, with a list of optimization Rules and input statistics.
+  """
+
+  rules: List[Rule]
+
+  def optimize(plan: PhysicalPlan, statistics) -> PhysicalPlan:
+    for rule in rules:
+      plan = rule.apply(plan, statistics)
+    return plan 
+
+class Rule():
+  """Rule to optimize plan.
+
+  A rule to optimize logical or physical plan. The rule should be cost-based and generate the plan with minimal cost.  
+  """
+  
+  def apply(plan: Plan, statistics) -> Plan:
+    # specific optimization logic in each rule
+```
+
+##### 2.2.2. Milestones
+
+* Step1: Introduce `Optimizer` class.
```

Comment:
At this stage, we won't have any optimization rules based on run-time stats, right? Might be good to:
1. Simplify the proposal to remove anything that is only needed for run-time optimization.
2. Add a simple `Rule` code example to demonstrate interface and functionality.

Reply:
> Simplify the proposal to remove anything that is only needed for run-time optimization.

Sure, removed. We probably still need input statistics though, so keep input statistics only.

> Add a simple Rule code example to demonstrate interface and functionality.

Added a simple pseudo code example for limit push down rule.